test/csr/csr.cc

```diff
@@ -0,0 +1,259 @@
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrw zero, %0, %1" ::"I"( CSR ), "r"( arg ) );
+      }
+    } else {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrw %0, %1, zero" : "=r"( old ) : "I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrw %0, %1, %2" : "=r"( old ) : "I"( CSR ), "r"( arg ) );
+      }
+    }
+    return old;
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Set, OpKind::Reg, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrs zero, %0, zero" ::"I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrs zero, %0, %1" ::"I"( CSR ), "r"( arg ) );
+      }
+      return old;
+    } else {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrs %0, %1, zero" : "=r"( old ) : "I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrs %0, %1, %2" : "=r"( old ) : "I"( CSR ), "r"( arg ) );
+      }
+      return old;
+    }
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Clear, OpKind::Reg, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrc zero, %0, zero" ::"I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrc zero, %0, %1" ::"I"( CSR ), "r"( arg ) );
+      }
+    } else {
+      if constexpr( ARG == 0 ) {
+        asm volatile( "csrrc %0, %1, zero" : "=r"( old ) : "I"( CSR ) );
+      } else {
+        size_t arg = ARG;
+        asm volatile( "csrrc %0, %1, %2" : "=r"( old ) : "I"( CSR ), "r"( arg ) );
+      }
+    }
+    return old;
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Write, OpKind::Imm, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      asm volatile( "csrrwi zero, %0, %1" ::"I"( CSR ), "K"( ARG ) );
+    } else {
+      asm volatile( "csrrwi %0, %1, %2" : "=r"( old ) : "I"( CSR ), "K"( ARG ) );
+    }
+    return old;
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Set, OpKind::Imm, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      asm volatile( "csrrsi zero, %0, %1" ::"I"( CSR ), "K"( ARG ) );
+    } else {
+      asm volatile( "csrrsi %0, %1, %2" : "=r"( old ) : "I"( CSR ), "K"( ARG ) );
+    }
+    return old;
+  }
+};
+
+template<uint16_t CSR, size_t ARG, bool RD0>
+struct CSRInst<CSROp::Clear, OpKind::Imm, CSR, ARG, RD0> {
+  size_t operator()() {
+    size_t old = 0;
+    if constexpr( RD0 ) {
+      asm volatile( "csrrci zero, %0, %1" ::"I"( CSR ), "K"( ARG ) );
+    } else {
+      asm volatile( "csrrci %0, %1, %2" : "=r"( old ) : "I"( CSR ), "K"( ARG ) );
+    }
+    return old;
+  }
+};
```

Comment:
When the operand to the CSR instruction is an immediate, we pass it using the RISC-V inline assembly modifier `K`, which indicates a 5-bit zero-extended immediate instead of a register. Like the register operand CSR instructions, when `rd == x0`  we explicitly code `x0` instead of using a local register.